lib/spack/spack/environment/environment.py

```diff
@@ -1341,15 +1341,36 @@ def concretize(self, force=False, tests=False):
         msg = "concretization strategy not implemented [{0}]"
         raise SpackEnvironmentError(msg.format(self.unify))
 
-    def deconcretize(self, spec):
+    def deconcretize(self, spec, concrete=True):
+        # If concrete, find all instances of concrete spec
+        # Else, find single instance of user spec
+        # the distinction is only relevant if multiple roots concretize
+        # to the same spec
+
         # spec has to be a root of the environment
-        index = self.concretized_user_specs.index(spec)
-        dag_hash = self.concretized_order.pop(index)
-        del self.concretized_user_specs[index]
+        if concrete:
+            dag_hash = spec.dag_hash()
+            indices = [
+                i
+                for i in range(len(self.concretized_order))
+                if self.concretized_order[i] == dag_hash
+            ]
+
+            for index in indices:
+                del self.concretized_order[index]
+                del self.concretized_user_specs[index]
```

Comment:
This isn't correct, right? Each `del` invalidates later indices b/c it modifies the list... is there a test for deconcretizing something with two concrete matches?

```suggestion
            filtered = [
                spec, h
                for spec, h in zip(self.concretized_user_specs, self.concretized_order)
                if h != dag_hash
            ]
            self.concretized_user_specs, self.concretized_order = zip(*filtered)
```